rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+start-date: 2023-07-21
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Require pull requests for all Nixpkgs commits.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently 197 people with commit access to Nixpkgs, all of whom can push directly to any branch without a pull request.
+Such pushes generally do not notify anybody and do not trigger CI.
+
+This makes such commits susceptible to:
+- Be anonymous
+- Include malicious code
+- Be broken
+- Have poor code quality
+
+While requiring pull requests isn't a panacea, it does help by:
+- Notifying [code owners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)
+- Running CI (though there's no requirement for it to succeed, see [future work][future])
+- Tying commits to a GitHub account
+- Being discoverable in the pull request list
+
+# Detailed design
+[design]: #detailed-design
+
+Turn on GitHub's "Require a pull request before merging" [branch protection rule](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule#creating-a-branch-protection-rule) for all branches whose commits get propagated into channels.
+This includes:
+- `master`: Used for unstable channels and branched into new release branches
+- `release-*`: Used for stable channels
+
+Staging branches are intentionally not included, because they will already require a pull request when they inevitably need to get merged into one of the above branches.
+The same applies to similar long-term branches like `haskell-packages`.
+
+A NixOS GitHub organization owner needs to implement this change and should therefore review this proposal.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Direct pushes listing
+
+Out of the 112217 commits to master in the last year[^1], _at most_ 58 (0.0517%) of them were direct pushes.
+
+[^1]: Unix epoch 1658361600 to 1689897600
+
+To determine whether a commit was pushed directly, the GitHub API was queried for pull requests associated with that commit (see [`associatedPullRequests`](https://docs.github.com/en/graphql/reference/objects#commit)).
+If this list includes a merged pull request to the Nixpkgs master branch, the commit is known to be merged with a pull request.
+Otherwise the commit could be directly pushed or be a false positives, which is why the above count is only an upper bound.
+All obvious false positives ([example](https://github.com/NixOS/nixpkgs/commit/b09d18903c24b8aca88100df86aa2fdd5f05dfcd)) have been removed from this count and listing already.
```

Comment:
Actually, have you kept count _how many_ obvious false positives you had to remove? To estimate how many non-obvious false positives we probably have remaining…

Reply:
There were 79 obvious false positive, which after another look are all from seemingly buggy merges using GitHub's web interface. Here's a complete listing:

<details>
<summary>Complete listing of false positives</summary>

- [5cd9b4264517](https://github.com/NixOS/nixpkgs/commit/5cd9b4264517ffba734556969d9d25ebeea55310) [@web-flow](https://github.com/web-flow) - Merge pull request #220739 from spacefrogg/factor-fix
- [8c619a1f3ced](https://github.com/NixOS/nixpkgs/commit/8c619a1f3cedd16ea172146e30645e703d21bfc1) [@web-flow](https://github.com/web-flow) - Merge pull request #215956 from K900/kf5-5.103.0
- [ae9b433138cf](https://github.com/NixOS/nixpkgs/commit/ae9b433138cf03df29af811bdec9ab6dd98cd825) [@web-flow](https://github.com/web-flow) - Merge pull request #215786 from wegank/qbittorrent-nox-darwin
- [7a0a5c995005](https://github.com/NixOS/nixpkgs/commit/7a0a5c99500579302e89407753878a38b692f61e) [@web-flow](https://github.com/web-flow) - Merge pull request #215759 from melvyn2/patch-4
- [c98696fbfc2f](https://github.com/NixOS/nixpkgs/commit/c98696fbfc2f2616c370ae36fe986029d97ef1dc) [@web-flow](https://github.com/web-flow) - Merge pull request #215744 from alyssais/kguiaddons-homepage
- [272017b7536d](https://github.com/NixOS/nixpkgs/commit/272017b7536deb5735def9f9b4bd0931188ae7fe) [@web-flow](https://github.com/web-flow) - Merge pull request #215743 from alyssais/extra-cmake-modules-homepage
- [4ba33a758f25](https://github.com/NixOS/nixpkgs/commit/4ba33a758f25950e8932b25b708bc37fd221825e) [@web-flow](https://github.com/web-flow) - Merge pull request #215489 from Alper-Celik/master
- [1e76265b3b56](https://github.com/NixOS/nixpkgs/commit/1e76265b3b561849147b65d710b22112dfe49efa) [@web-flow](https://github.com/web-flow) - Merge pull request #215177 from Emantor/bump/xwayland
- [27cde435fe6a](https://github.com/NixOS/nixpkgs/commit/27cde435fe6a6749fcf54df849e19fe7ef9c9787) [@web-flow](https://github.com/web-flow) - Merge pull request #214857 from dotlambda/kbibtex-0.9.3.1
- [42b3a6354fde](https://github.com/NixOS/nixpkgs/commit/42b3a6354fde25436733f56eebcd17d3fdf857a3) [@web-flow](https://github.com/web-flow) - Merge pull request #206559 from doronbehar/pkg/musescore
- [b594c9af3621](https://github.com/NixOS/nixpkgs/commit/b594c9af362189f745b73c0092b3d6a3407f63a5) [@web-flow](https://github.com/web-flow) - Merge pull request #214552 from r-ryantm/auto-update/kstars
- [ffbc5e161a5a](https://github.com/NixOS/nixpkgs/commit/ffbc5e161a5a089d6feb29b97f36b68c760de500) [@web-flow](https://github.com/web-flow) - Merge pull request #214220 from K900/kde-rounded-corners-0.3.0
- [c2947b5459c3](https://github.com/NixOS/nixpkgs/commit/c2947b5459c340da13c54efe825f4c56f06114e4) [@web-flow](https://github.com/web-flow) - Merge pull request #214169 from K900/kde-gear-22.12.2
- [e14a364a736b](https://github.com/NixOS/nixpkgs/commit/e14a364a736beb30fd5a311aa5889a7b2bc9b129) [@web-flow](https://github.com/web-flow) - Merge pull request #214005 from r-ryantm/auto-update/umockdev
- [3da3678c1719](https://github.com/NixOS/nixpkgs/commit/3da3678c17197e5373cfc5ad3b0ccd5f570f001d) [@web-flow](https://github.com/web-flow) - Merge pull request #213413 from K900/plasma-wayland-protocols-1.10.0
- [0e4512f10180](https://github.com/NixOS/nixpkgs/commit/0e4512f10180955a0feb584248c6892e6a1ac891) [@web-flow](https://github.com/web-flow) - Merge pull request #213099 from r-ryantm/auto-update/haruna
- [30e84e2f8175](https://github.com/NixOS/nixpkgs/commit/30e84e2f8175dadc8e53ec389e0d8c06277cedb9) [@web-flow](https://github.com/web-flow) - Merge pull request #206930 from brcha/davinci_resolve_fix_qt_plugins
- [c5b2c8390c15](https://github.com/NixOS/nixpkgs/commit/c5b2c8390c1506cfd8899eaa3f4ea165d3c9f97d) [@web-flow](https://github.com/web-flow) - Merge pull request #212802 from Zimmi48/update-copilot
- [437b1938bcd1](https://github.com/NixOS/nixpkgs/commit/437b1938bcd1546629b2e88ce676ceb98a31a8c6) [@web-flow](https://github.com/web-flow) - Merge pull request #212611 from r-ryantm/auto-update/goda
- [4be063c1732e](https://github.com/NixOS/nixpkgs/commit/4be063c1732e919b5788e862aa224fbe351bbf25) [@web-flow](https://github.com/web-flow) - Merge pull request #212481 from qowoz/fzf
- [542ef388022f](https://github.com/NixOS/nixpkgs/commit/542ef388022ff00ef74c1d5f75af81426dc2ea63) [@web-flow](https://github.com/web-flow) - Merge pull request #211602 from Jayman2000/ecwolf-1.4.0
- [15eb39e56f67](https://github.com/NixOS/nixpkgs/commit/15eb39e56f67a70ca3628ebf03c0c3c55c41aa35) [@web-flow](https://github.com/web-flow) - Merge pull request #209224 from wegank/qbittorrent-darwin
- [7d81cb9db363](https://github.com/NixOS/nixpkgs/commit/7d81cb9db363aaa9bf8500699352c0812b6ff065) [@web-flow](https://github.com/web-flow) - Merge pull request #210542 from michaelBelsanti/catppuccin-kde
- [78a377ed2880](https://github.com/NixOS/nixpkgs/commit/78a377ed2880bcd0f5a4fc8adc7f8e38c2c3a5a2) [@web-flow](https://github.com/web-flow) - Merge pull request #210290 from NickCao/qt642
- [d7021b0c4d62](https://github.com/NixOS/nixpkgs/commit/d7021b0c4d62f4792ebfffb07f76b49b51595eda) [@web-flow](https://github.com/web-flow) - Merge pull request #210709 from K900/kf5-5.102.0
- [5c2347a3e961](https://github.com/NixOS/nixpkgs/commit/5c2347a3e961a3dd30f4e8f3831738c787a3c497) [@web-flow](https://github.com/web-flow) - Merge pull request #210581 from r-ryantm/auto-update/kid3
- [c96b17f7675e](https://github.com/NixOS/nixpkgs/commit/c96b17f7675ed4b899cbf5927f127f7d272a9569) [@web-flow](https://github.com/web-flow) - Merge pull request #209439 from K900/kde-gear-22.12.1
- [505955711ae9](https://github.com/NixOS/nixpkgs/commit/505955711ae93ccc326a2ed0d9c55a9f7c42fe29) [@web-flow](https://github.com/web-flow) - Merge pull request #209433 from K900/plasma-5.26.5
- [03287dc59d23](https://github.com/NixOS/nixpkgs/commit/03287dc59d2363061e4f1f697d5c05ac67c7c94a) [@web-flow](https://github.com/web-flow) - Merge pull request #209019 from jojosch/haruna-update
- [81271652095d](https://github.com/NixOS/nixpkgs/commit/81271652095dc071bd6c04e393af6af237842f68) [@web-flow](https://github.com/web-flow) - minio: add legacy fs version 2022-10-24T18-35-07Z (#206376)
- [b09d18903c24](https://github.com/NixOS/nixpkgs/commit/b09d18903c24b8aca88100df86aa2fdd5f05dfcd) [@web-flow](https://github.com/web-flow) - Merge pull request #205921 from ulrikstrid/ulrikstrid--kde-frameworks_5-101
- [5b856c23929b](https://github.com/NixOS/nixpkgs/commit/5b856c23929ba03d25cac9bb1efe1ceb21a0da2d) [@web-flow](https://github.com/web-flow) - Merge pull request #205625 from r-ryantm/auto-update/syncthingtray
- [4da241459471](https://github.com/NixOS/nixpkgs/commit/4da241459471020a1f0d2ee6ed9e525f67a9122c) [@web-flow](https://github.com/web-flow) - Merge pull request #203368 from maralorn/nom
- [f6a53df3df6d](https://github.com/NixOS/nixpkgs/commit/f6a53df3df6dc850f214c5e0a72cbbfa9adbb465) [@web-flow](https://github.com/web-flow) - Merge pull request #201808 from yaxitech/pulumi-3.47.0
- [3c0f6f555b7e](https://github.com/NixOS/nixpkgs/commit/3c0f6f555b7ecc12154e1e47ba77d5aaf97e863a) [@web-flow](https://github.com/web-flow) - Merge pull request #201459 from tjni/od
- [5cb03b3f5a1b](https://github.com/NixOS/nixpkgs/commit/5cb03b3f5a1b0556a79aa34f0ddcb053fd02e176) [@web-flow](https://github.com/web-flow) - Merge pull request #201460 from trofi/libagar-without-xlibsWrapper
- [fc8a4f729105](https://github.com/NixOS/nixpkgs/commit/fc8a4f72910595b287fdde96b66f75fec760715f) [@web-flow](https://github.com/web-flow) - Merge pull request #201521 from hmenke/alps
- [9727e904e901](https://github.com/NixOS/nixpkgs/commit/9727e904e9010803924005fd9ff197de0ffa9b4d) [@web-flow](https://github.com/web-flow) - Merge pull request #201469 from hmenke/git-branchless
- [3b0b1be8784f](https://github.com/NixOS/nixpkgs/commit/3b0b1be8784fe6676945f69033e6c2bba26fa911) [@web-flow](https://github.com/web-flow) - Merge pull request #201679 from r-ryantm/auto-update/brev-cli
- [efa60e11cf4f](https://github.com/NixOS/nixpkgs/commit/efa60e11cf4f5e8f6c2857616057f71875ee020e) [@web-flow](https://github.com/web-flow) - Merge pull request #201535 from r-ryantm/auto-update/picom
- [639550cc2ee6](https://github.com/NixOS/nixpkgs/commit/639550cc2ee630320b4c5029f09c028e20abe94a) [@web-flow](https://github.com/web-flow) - matcha-gtk-theme: 2022-06-07 -> 2022-11-15 (#201410)
- [3e1a3aa9392f](https://github.com/NixOS/nixpkgs/commit/3e1a3aa9392fd2be39717eb1ea2a08253d1ba3fd) [@web-flow](https://github.com/web-flow) - Merge pull request #201376 from whentze/master
- [bc5ad9707805](https://github.com/NixOS/nixpkgs/commit/bc5ad970780580ec272b3a6bc5f9bddf4e250d3c) [@web-flow](https://github.com/web-flow) - Merge pull request #200097 from efx/fluent-bit-support-postgresql-out
- [43a2c2f1a0cc](https://github.com/NixOS/nixpkgs/commit/43a2c2f1a0cc28c1ffe9afbe2708ac6d49a8c47a) [@web-flow](https://github.com/web-flow) - Merge pull request #200155 from r-ryantm/auto-update/python310Packages.mastodon-py
- [25528d7ae1d5](https://github.com/NixOS/nixpkgs/commit/25528d7ae1d576b25b066be30c2ebcfa8bb28c59) [@web-flow](https://github.com/web-flow) - Merge pull request #199990 from prusnak/electron
- [5f24d55ae4f2](https://github.com/NixOS/nixpkgs/commit/5f24d55ae4f27f0134f6db299fff37efc1711a89) [@web-flow](https://github.com/web-flow) - Merge pull request #199700 from fabaff/plugwise-bump
- [9e7feda53c23](https://github.com/NixOS/nixpkgs/commit/9e7feda53c23bd24a5c40874b242b6a0ef735a2e) [@web-flow](https://github.com/web-flow) - Merge pull request #199153 from r-ryantm/auto-update/babashka
- [96d7f290878d](https://github.com/NixOS/nixpkgs/commit/96d7f290878dec43e2688940467d24499c208cd9) [@web-flow](https://github.com/web-flow) - Merge pull request #176553 from MalteT/fix/smartctl-exporter-override
- [220943e6d6cd](https://github.com/NixOS/nixpkgs/commit/220943e6d6cdd15136da628bdb74dc06660e19b0) [@web-flow](https://github.com/web-flow) - Merge pull request #197427 from r-ryantm/auto-update/libtorrent-rasterbar-2_0_x
- [0d1d1f9ac121](https://github.com/NixOS/nixpkgs/commit/0d1d1f9ac121a034d9a0bd3cc96a14531a441d1e) [@web-flow](https://github.com/web-flow) - Merge pull request #197426 from r-ryantm/auto-update/libtorrent-rasterbar
- [46e67960e03e](https://github.com/NixOS/nixpkgs/commit/46e67960e03e98f5e62a25c3c6ca5fc466f5004e) [@web-flow](https://github.com/web-flow) - Merge pull request #197279 from fabaff/metasploit-bump
- [0d37b5bf11ca](https://github.com/NixOS/nixpkgs/commit/0d37b5bf11caf3850d9d060237848451c576db86) [@web-flow](https://github.com/web-flow) - Merge pull request #196554 from r-ryantm/auto-update/autorestic
- [3bcc8e1ac7ec](https://github.com/NixOS/nixpkgs/commit/3bcc8e1ac7ec6040219acc8da4cda6df394cddb1) [@web-flow](https://github.com/web-flow) - Merge pull request #196408 from mweinelt/thunderbird
- [3367189eb6f6](https://github.com/NixOS/nixpkgs/commit/3367189eb6f612eea649588bb217856c393b82e7) [@web-flow](https://github.com/web-flow) - Merge pull request #196461 from r-ryantm/auto-update/lefthook
- [499e75fe3194](https://github.com/NixOS/nixpkgs/commit/499e75fe319408faefa77b480fac3e44ed259f65) [@web-flow](https://github.com/web-flow) - Merge pull request #189406 from r-ryantm/auto-update/qbittorrent
- [5f41d9525a9e](https://github.com/NixOS/nixpkgs/commit/5f41d9525a9e6c6d665e8e7d172d5ebac6a60df1) [@web-flow](https://github.com/web-flow) - Merge pull request #195044 from r-ryantm/auto-update/python310Packages.msal
- [898dd0e87266](https://github.com/NixOS/nixpkgs/commit/898dd0e87266ae65d85511505cc5bf2166ebcd91) [@web-flow](https://github.com/web-flow) - Merge pull request #194225 from figsoda/artem
- [9269709bec17](https://github.com/NixOS/nixpkgs/commit/9269709bec17a62e3bc66d23006055ef96ef1878) [@web-flow](https://github.com/web-flow) - Merge pull request #193330 from r-ryantm/auto-update/python310Packages.types-urllib3
- [1e8213d556ea](https://github.com/NixOS/nixpkgs/commit/1e8213d556eafca5c10b56ccf94a94bb581fcdcc) [@web-flow](https://github.com/web-flow) - Merge pull request #191306 from r-ryantm/auto-update/flyway
- [e221ed41ec4a](https://github.com/NixOS/nixpkgs/commit/e221ed41ec4a93b6419bd923a96d7592589bd6ce) [@web-flow](https://github.com/web-flow) - Merge pull request #190680 from anthonyroussel/fix-flask-wtf
- [7d9aaa68da8e](https://github.com/NixOS/nixpkgs/commit/7d9aaa68da8e32798e42b5b399950951bf59ecf9) [@web-flow](https://github.com/web-flow) - Merge pull request #190053 from anthonyroussel/fix-mitmproxy
- [aaadf6612a3d](https://github.com/NixOS/nixpkgs/commit/aaadf6612a3db91ededa305978dd993520aa9110) [@web-flow](https://github.com/web-flow) - Merge pull request #189116 from stigtsp/fix/perl-io-async-ssl-upgrade-error-staging-next
- [9cdfb3d950b5](https://github.com/NixOS/nixpkgs/commit/9cdfb3d950b5f83efe3a560b915729e3328f88bf) [@web-flow](https://github.com/web-flow) - Merge pull request #188804 from gador/flask-login-0.6.2
- [56daf120b113](https://github.com/NixOS/nixpkgs/commit/56daf120b113f6e9d8995bf0b73fca6d7c1d1702) [@web-flow](https://github.com/web-flow) - Merge pull request #187121 from alyssais/colord-musl
- [9fdef95e3ef4](https://github.com/NixOS/nixpkgs/commit/9fdef95e3ef49944ad2d971be3867ac35cf793d5) [@web-flow](https://github.com/web-flow) - Merge pull request #187531 from tjni/fix-bitstring
- [26c03f3bf15f](https://github.com/NixOS/nixpkgs/commit/26c03f3bf15f148765ee1c071e20ba135b3ba108) [@web-flow](https://github.com/web-flow) - Merge pull request #187590 from trofi/libgccjit-lib-stripping-fix
- [07b19b321028](https://github.com/NixOS/nixpkgs/commit/07b19b321028b2852e828475b60cd6e15b1e344a) [@web-flow](https://github.com/web-flow) - Merge pull request #187200 from LibreCybernetics/babl-fix
- [318695815990](https://github.com/NixOS/nixpkgs/commit/3186958159908a0d4e481c5ef386847525671e4a) [@web-flow](https://github.com/web-flow) - Merge pull request #186580 from tjni/orjson-tzdata
- [ff72265385eb](https://github.com/NixOS/nixpkgs/commit/ff72265385ebc3615fb25d41c90b2cb945106c01) [@web-flow](https://github.com/web-flow) - Merge pull request #186842 from Lassulus/django-archivebox-hash
- [775bb14b9147](https://github.com/NixOS/nixpkgs/commit/775bb14b914742f47e8ba288a6ade4e29db22c98) [@web-flow](https://github.com/web-flow) - Merge pull request #186555 from tjni/fix-loguru
- [b9106e11154e](https://github.com/NixOS/nixpkgs/commit/b9106e11154e264a0b964649f9ba72f4898edace) [@web-flow](https://github.com/web-flow) - Merge pull request #186212 from stigtsp/perl-5.36.0-fallout-staging-next
- [58547ae85a10](https://github.com/NixOS/nixpkgs/commit/58547ae85a10855beffc33694443ea8fb6c36a90) [@web-flow](https://github.com/web-flow) - Merge pull request #184902 from trofi/fix-openldap-markdown
- [7be3a05eb8f1](https://github.com/NixOS/nixpkgs/commit/7be3a05eb8f1e4aadb6fa11d886bbfa5c352d1e6) [@web-flow](https://github.com/web-flow) - Merge pull request #182834 from mayflower/security-fixes
- [bbe75628ca71](https://github.com/NixOS/nixpkgs/commit/bbe75628ca71378a02891bb5f5ecb769832da3bf) [@web-flow](https://github.com/web-flow) - Merge pull request #182692 from mweinelt/uvicorn
- [35875ee05506](https://github.com/NixOS/nixpkgs/commit/35875ee05506b0a55395ea3f7a9bb11b104ed383) [@web-flow](https://github.com/web-flow) - Merge pull request #182727 from trofi/raptor2-dynamic-by-default
- [66f6d190543e](https://github.com/NixOS/nixpkgs/commit/66f6d190543e7e23f3d09521246dc298f8a38cac) [@web-flow](https://github.com/web-flow) - Merge pull request #182687 from YorikSar/upower-bump
- [f4d202194fb3](https://github.com/NixOS/nixpkgs/commit/f4d202194fb3cb10dd738d96006e8371fd195132) [@web-flow](https://github.com/web-flow) - Merge pull request #182668 from YorikSar/umockdev-bump
- [f315837ee425](https://github.com/NixOS/nixpkgs/commit/f315837ee4259e6e604735e4a9dc460e9a1dbaf3) [@web-flow](https://github.com/web-flow) - Merge pull request #182377 from trofi/fix-sbcl-on-aarch64-darwin
- [84bdacd8cb21](https://github.com/NixOS/nixpkgs/commit/84bdacd8cb211316bbf694061b2e6005c5f45e5d) [@web-flow](https://github.com/web-flow) - Merge pull request #182399 from SuperSandro2000/eventlet

</details>

Though I don't think this needs to be part of the RFC because it doesn't change any arguments.

Reply:
Well, if you just add 79 to show that _clear_ false positives are more numerours than «maybe, who knows», it does strengthen the argument about rarity of true directy pushes and their removal changing nothing at all.

Reply:
Hmm I don't think it matters though, it's all just caused by imperfect ways to get at the right data (my script and GitHub bugs in this case). These false positives have no influence on any arguments and in fact only distract from the main point. In fact I should just "fix" my script to not count any commits from @web-flow at all, then I don't need to mention this at all because there aren't any false positives anymore.

Note that these false positives in my script are completely unrelated to the false positives in the automatic notifications in https://github.com/NixOS/nixpkgs/issues/118661, which _do_ have an impact because they annoy people.

Reply:
The false positives are relevant for two reasons. 

1. We know that what this RFC seeks to forbid is rare, but we don't even know if it ever happens at all nowadays.
2. We do not have a real option to allow but closely watch and check after the fact if the case was exceptional, because no amount of waiting before asking GitHub will give you good data. «Allow-but-monitor» could be listed as an a priori plausible alternative by the way (with these false positives explaining why it doesn't work).